### PR TITLE
Detect font smoothing

### PR DESCRIPTION
**Pull request by sndrs**
We're still getting reports of fonts rendering terribly in some circumstances on windows - seemingly edge cases where font-smoothing is not/cannot be enabled e.g. this is Windows 7 for a user on citrix virtual desktop:

b8hwf0xiiaacazd

The thrust of this PR is to manually detect font-smoothing on Windows desktops, in non-IE browsers only, and render using only system fallbacks if it's disabled:
- disabling font-smoothing in IE is [long-winded](http://stackoverflow.com/questions/5427315/disable-cleartype-text-anti-aliasing-in-ie9#tab-top) (it ignores system-level settings). If you want to go that far, this won't support it. 
- chrome/ff observe system-level settings, including in win 7, so it's not enough to simply serve auto-hinted fonts.

It works by drawing a `@` character to a canvas element, and inspecting it for partially opaque pixels (derived from Zoltan Hawryluk's [original idea](http://www.useragentman.com/blog/2009/11/29/how-to-detect-font-smoothing-using-javascript)) - any non-black/transparent pixel and you're smoothing.
- it's checked once every 30 days if we think you have smoothing on. If you disable fonts manually, or we decide we need to disable them on your behalf, they're disabled for a year and the smoothing check doesn't happen. 
- you can force a test (if your browser is one we test) using a `#check-smoothing` hash fragment

I've very much felt my way through this, so please go to town on it...

cc @phamann @paperboyo @stephanfowler @crifmulholland 
# To test
## Windows where font smoothing is off
- [x] GU_fonts cookie value should = "off"
### In XP
- [x] Chrome 
- [x] Chrome - Incognito mode
- [x] Firefox - issue - dropping GU_fonts cookie with value set to "on"
- [x] IE8 - does its own thing and ignores our fonts anyway
### In Windows 7
- [x] Chrome - with font smoothing off - cookie still showing fonts = true
- [x] Firefox - with font smoothing off - cookie still showing fonts = true
- [x] IE9
## Windows pcs where font smoothing is active
- [x] GU_fonts cookie value should = "on"
### In XP
- [x] Chrome 
- [x] Chrome - incognito mode
- [x] Firefox
- [x] IE8
### In Windows 7
- [x] Chrome
- [x] Firefox
- [x] IE9
### On OSX
- [x] Chrome
- [x] Firefox
- [x] Safari
### Other
- [x] iPadAir/mini
- [x] iPhone4S
- [x] Android (Galaxy S3)

### CODE REVIEW

File: common/app/views/fragments/javaScriptLaterSteps.scala.html [hunk 1]
```diff
@@ -60,61 +64,158 @@
...
         @if(FontSwitch.isSwitchedOn) {
+            // Check to see if you should get webfonts, and then try to load them from localStorage if so
+            var cookieData = 'GU_fonts=off; domain=.theguardian.com; path=/';
+
+            function disableFonts () {
+                document.cookie = cookieData + '; max-age=' + (60 * 60 * 24 * 365);
+            }
+
+            function enableFonts () {
+                document.cookie = cookieData + '; expires=Thu, 01 Jan 1970 00:00:00 GMT';
```
Comment (phamann) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: :cookie: :goberserk:

File: common/app/views/fragments/javaScriptLaterSteps.scala.html [hunk 2]
```diff
@@ -60,61 +64,158 @@
...
+                            nameAndCacheKey = font.getAttribute(dataAttrName).match(/fonts\/([^/]*?)\/?([^/]*)\.(woff2|woff|tff).json$/),
+                            fontData = localStorage['gu.fonts.' + nameAndCacheKey[2] + '.' + nameAndCacheKey[1]];
+
+                        if (fontData) {
+                            font.innerHTML = JSON.parse(fontData).value;
+                            font.setAttribute('data-loaded-from', 'local');
+                        }
+                    }
+                }
+            }
```
Comment (phamann) [inline common/app/views/fragments/javaScriptLaterSteps.scala.html]: Why have you used this function here?

**Comment by phamann** [comment]
:+1: Looks ok to me. But lets have a chat in person before you ship, specifically about:
- use of cookies
- duplication of code
- I can understand the logic and intention and bit more clearly :smile:

**Comment by robertberry** [comment]
Nuts. If it works, it works though :+1:

**Comment by michaelwmcnamara** [comment]
check display on windows pcs where font smoothing is off
           -[ ] GU_fonts cookie value should = "off"
           -Check display of fonts
           -[ ] XP
                  -[ ] Chrome
                  -[ ] Firefox
                  -[ ] IE8
           -[ ] Windows 7
                  -[ ] Chrome
                  -[ ] Firefox
                  -[ ] IE9

check display on windows pcs where font smoothing is active
           -[ ] GU_fonts cookie value should = "on"
           -Check display of fonts
           -[ ] XP
                  -[ ] Chrome
                  -[ ] Firefox
                  -[ ] IE8
           -[ ] Windows 7
                  -[ ] Chrome
                  -[ ] Firefox
                  -[ ] IE9

Check display of fonts on other devices is unchanged
            -[ ] Mac
                  -[ ] Chrome
                  -[ ] Firefox
                  -[ ] Safari
             -[ ] iPadAir/mini
             -[ ] iphone4S
             -[ ] android (Galaxy S3)

**Comment by michaelwmcnamara** [comment]
Windows 7 font smooting cookie dropped and set to true even when font smoothing is switched off.

**Comment by phamann** [comment]
Whats the status of this?

**Comment by sndrs** [comment]
micahel found prob with win7, looking at it

On Mon Feb 09 2015 at 11:00:14 AM Patrick Hamann <notifications@github.com>
wrote:

> Whats the status of this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/frontend/pull/8146#issuecomment-73490877>.
>
> ------------------------------
> Visit theguardian.com. On your mobile and tablet, download the Guardian
> iPhone and Android apps theguardian.com/guardianapp and our tablet
> editions theguardian.com/editions.  Save up to 57% by subscribing to the
> Guardian and Observer - choose the papers you want and get full digital
> access.  Visit subscribe.theguardian.com
>
> This e-mail and all attachments are confidential and may also be
> privileged. If you are not the named recipient, please notify the sender
> and delete the e-mail and all attachments immediately. Do not disclose the
> contents to another person. You may not use the information for any
> purpose, or store, or copy, it in any way.  Guardian News & Media Limited
> is not liable for any computer viruses or other material transmitted with
> or as part of this e-mail. You should employ virus checking software.
>
> Guardian News & Media Limited is a member of Guardian Media Group plc. Registered
> Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered
> in England Number 908396
>
>
>

-- 

------------------------------
Visit theguardian.com. On your mobile and tablet, download the Guardian 
iPhone and Android apps theguardian.com/guardianapp and our tablet editions 
theguardian.com/editions.  Save up to 57% by subscribing to the Guardian 
and Observer - choose the papers you want and get full digital access. 
 Visit subscribe.theguardian.com

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

**Comment by michaelwmcnamara** [comment]
All looking good now - ready to ship.

**Pull request merged**